File: Rosu/programmazione/lezioni/java/mutable-round-table/src/ro/razvan/uniud/mutableRoundTable/java/RoundTable.java
package ro.razvan.uniud.mutableRoundTable.java;

import org.jetbrains.annotations.NotNull;
import ro.razvan.uniud.intSList.IntSList;

class RoundTable {

    @NotNull
    private IntSList knightsLeft;

    @NotNull
    private IntSList knightsStanding;

    private boolean served;

    RoundTable(final int n) {
        knightsLeft = rangeIter(1, n);
        knightsStanding = IntSList.Companion.getNULL_INT_S_LIST();
    }

    private static IntSList rangeIter(final int start, int end) {
        var list = IntSList.Companion.getNULL_INT_S_LIST();

        while (end >= start) {
            list = list.cons(end);
            end--;
        }

        return list;
    }

    private static IntSList rangeRec(final int start, final int end) {
        if (start > end) {
            return IntSList.Companion.getNULL_INT_S_LIST();
        } else {
            return new IntSList(start, IntSList.Companion.getNULL_INT_S_LIST()).append(rangeRec(start + 1, end));
        }
    }

    static int count(int n) {

        final var roundTable = new RoundTable(n);

        while (n-- > 1) {
            roundTable.serveKnightAndPassPitcher();
        }

        return roundTable.getKnightWithPitcher();
    }

    int getKnightWithPitcher() {
        return knightsLeft.getCar();
    }

    void serveKnight() {
        knightsStanding = knightsStanding.cons(knightsLeft.getCdr().getCar());
        knightsLeft = knightsLeft.getCdr().getCdr().cons(knightsLeft.getCar());
        served = true;
    }

    void handPitcher() {
        knightsLeft = knightsLeft.getCdr().append(new IntSList(knightsLeft.getCar(), IntSList.Companion.getNULL_INT_S_LIST()));
        served = false;
    }

    void serveKnightAndPassPitcher() {
        knightsStanding = knightsStanding.cons(knightsLeft.getCdr().getCar());
        knightsLeft = knightsLeft.getCdr().getCdr().append(new IntSList(knightsLeft.getCar(), IntSList.Companion.getNULL_INT_S_LIST()));
        served = false;
    }

}

File: Rosu/programmazione/lezioni/java/mutable-round-table/src/ro/razvan/uniud/mutableRoundTable/kotlin/RoundTable.kt
package ro.razvan.uniud.mutableRoundTable.kotlin

import ro.razvan.uniud.intSList.IntSList

class RoundTable {

    companion object {

        private fun rangeRec(start: Int, end: Int): IntSList {
            return if (start > end) {
                IntSList()
            } else {
                IntSList(start).append(rangeRec(start + 1, end))
            }
        }

        private fun rangeIter(start: Int, end: Int): IntSList {
            var list = IntSList()
            var i = end

            while (i >= start) {
                list = IntSList(i, list)
                i--
            }

            return list
        }

        fun count(n: Int): Int {
            val roundTable = RoundTable(n)
            var i = n
            while (i-- > 1) {
                roundTable.serveKnightAndHandPitcher()
            }

            return roundTable.knightWithPitcher
        }

    }

    var knightsLeft: IntSList
    private var knightsStanding: IntSList
    private var served: Boolean

    val knightServed: Boolean
        get() = served

    val knightWithPitcher: Int
        get() = knightsLeft.car

    val knightsAtTable: Int
        get() = knightsLeft.length

    constructor(n: Int) {
        knightsLeft = rangeIter(1, n)
        served = false
        knightsStanding = IntSList.NULL_INT_S_LIST
    }

    fun serveKnight() {
        knightsLeft = knightsLeft.cdr.cdr.cons(knightsLeft.car)
        knightsStanding = IntSList(knightsLeft.cdr.car, knightsStanding)
        served = true
    }

    fun handPitcher() {
        knightsLeft = knightsLeft.cdr.append(IntSList(knightsLeft.car))
        served = false
    }

    fun serveKnightAndHandPitcher() {
        knightsStanding = IntSList(knightsLeft.cdr.car, knightsStanding)
        knightsLeft = knightsLeft.cdr.cdr.append(IntSList(knightsLeft.car))
        served = false
    }


}